install_driver/tasks/ubuntu.yaml

```diff
@@ -8,46 +7,45 @@
   ansible.builtin.set_fact:
     major_driver_version: "{{ driver_version.split('.')[0] }}"
 
+- name: Register common CUDA driver packages
+  ansible.builtin.set_fact:
+    pkgs:
+      - libnvidia-cfg1
+      - libnvidia-compute
+      - libnvidia-decode
+      - libnvidia-encode
+      - libnvidia-extra
+      - libnvidia-fbc1
+      - libnvidia-gl
+      - nvidia-compute-utils
+      - nvidia-dkms
+      - nvidia-driver
+      - nvidia-kernel-source
+      - nvidia-utils
+      - xserver-xorg-video-nvidia
+
+- name: Append conditional CUDA driver packages (>=450,<=470)
+  when: (major_driver_version|int) >=450 and (major_driver_version|int) <= 470
+  ansible.builtin.set_fact:
+    pkgs: "{{ pkgs + ['libnvidia-ifr1'] }}"
+
 - name: Install cuda-keyring package (Ubuntu)
   ansible.builtin.apt:
     deb: "https://developer.download.nvidia.com/compute/cuda/repos/{{ distrib_repo_name }}/{{ arch_repo_name }}/cuda-keyring_1.0-1_all.deb"
 
 - name: Install driver (Ubuntu)
   ansible.builtin.apt:
-    allow_change_held_packages: yes
```

Comment:
Can we still upgrade packages that are marked as hold without this option? We need to be able to do this if we want to upgrade a node

Reply:
No. This is a tricky situation.

This feature isn't available in the `2.12.5` version of `ansible-core` [that Kubespray uses](https://github.com/kubernetes-sigs/kubespray/blob/release-2.20/requirements-2.12.txt#L2).

It was introduced in `2.13` ([src](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/apt_module.html#parameter-allow_change_held_packages)).

Even though the `ansible-core` version difference is only a minor version, that minor version bump causes further incompatibilities with the Kubespray versions of `jinja2` and `ansible` (the non-`core` package).

So if we tried to adopt `2.13`, we'd be incompatible with the rest of the Kubespray dependencies, which could cause issues.

I tried some other options including:

- Adding an `unhold` task that runs on all the packages before the `Install Driver` task.
  - the issue with this approach is that the `unhold` task doesn't accept wildcard versions (e.g. `nvidia-driver-*`). So when you try to upgrade a driver from `495` to `520`, this approach would attempt to `unhold` driver `520` (leaving the `495` driver still held) and then fail on the subsequent `Install driver` task
- Explicitly passing `force-hold` to the [dpkg_options](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/apt_module.html#parameter-dpkg_options) parameter.
  - this doesnt work due to [this issue](https://github.com/ansible/ansible/issues/72224)

In e73d730, I added a task that will `unhold` all the packages if the `unhold_driver_version` variable is defined.

I used this successfully like this:

```sh
ansible-playbook \
  -e unhold_driver_version=495 \
  --limit=k8s-node-7 \
  ansible/run.yaml
```

It's not ideal, but it's a minor inconvenience that allows us to continue using Kubespray's `requirements.txt` file.